pkg/skaffold/output/log/log.go

```diff
@@ -28,100 +30,170 @@ import (
+	// Commonly used for hooks to send errors to an error tracking service.
+	ErrorLevel
+	// WarnLevel level. Non-critical entries that deserve eyes.
+	WarnLevel
+	// InfoLevel level. General operational entries about what's going on inside the
+	// application.
+	InfoLevel
+	// DebugLevel level. Usually only enabled when debugging. Very verbose logging.
+	DebugLevel
+	// TraceLevel level. Designates finer-grained informational events than the Debug.
+	TraceLevel
+)
+
+// Level type for logging levels
+type Level uint32
+
+// AllLevels exposes all logging levels
+var AllLevels = []Level{
+	PanicLevel,
+	FatalLevel,
+	ErrorLevel,
+	WarnLevel,
+	InfoLevel,
+	DebugLevel,
+	TraceLevel,
+}
+
+// DefaultLogLevel for the global Skaffold logger
+const DefaultLogLevel = WarnLevel
 
 type contextKey struct{}
 
-var AllLevels = logrus.AllLevels
 var ContextKey = contextKey{}
 
+// logger is the global logrus.Logger for Skaffold
+// TODO: Make this not global.
+var logger = New()
+
 type EventContext struct {
 	Task    constants.Phase
 	Subtask string
 }
 
+// String converts the Level to a string. E.g. PanicLevel becomes "panic".
+func (level Level) String() string {
+	switch level {
+	case TraceLevel:
+		return "trace"
+	case DebugLevel:
+		return "debug"
+	case InfoLevel:
+		return "info"
+	case WarnLevel:
+		return "warning"
+	case ErrorLevel:
+		return "error"
+	case FatalLevel:
+		return "fatal"
+	case PanicLevel:
+		return "panic"
+	}
+	return "unknown"
+}
+
 // Entry takes an context.Context and constructs a logrus.Entry from it, adding
 // fields for task and subtask information
 func Entry(ctx context.Context) *logrus.Entry {
 	val := ctx.Value(ContextKey)
 	if eventContext, ok := val.(EventContext); ok {
-		return logrus.WithFields(logrus.Fields{
+		return logger.WithFields(logrus.Fields{
 			"task":    eventContext.Task,
 			"subtask": eventContext.Subtask,
 		})
 	}
 
 	// Use constants.DevLoop as the default task, as it's the highest level task we
 	// can default to if one isn't specified.
-	return logrus.WithFields(logrus.Fields{
+	return logger.WithFields(logrus.Fields{
 		"task":    constants.DevLoop,
 		"subtask": constants.SubtaskIDNone,
 	})
 }
 
 // IsDebugLevelEnabled returns true if debug level log is enabled.
 func IsDebugLevelEnabled() bool {
-	return logrus.IsLevelEnabled(logrus.DebugLevel)
+	return logger.IsLevelEnabled(logrus.DebugLevel)
 }
 
 // IsTraceLevelEnabled returns true if trace level log is enabled.
 func IsTraceLevelEnabled() bool {
-	return logrus.IsLevelEnabled(logrus.TraceLevel)
+	return logger.IsLevelEnabled(logrus.TraceLevel)
 }
 
+// New returns a new logrus.Logger.
+// We use a new instance instead of the default logrus singleton to avoid clashes with dependencies that also use logrus.
 func New() *logrus.Logger {
 	return logrus.New()
 }
 
 // KanikoLogLevel makes sure kaniko logs at least at Info level and at most Debug level (trace doesn't work with Kaniko)
 func KanikoLogLevel() logrus.Level {
```

Comment:
This function doesn't appear to be called anywhere, except for the unit test. I left the function signature unchanged, in case I missed something.

Reply:
Yeah, we've actually been using the global logrus logger, but we should definitely completely move off of that. I can make a follow up to this to have us use a local one that we own